docs/decisions/0040-OTel-semantic-convention.md

```diff
@@ -0,0 +1,628 @@
+    block:Clients
+        columns 3
+        ConnectorTypeClientA["Instrumented client SDK<br>(i.e. Azure OpenAI client)"]
+        ConnectorTypeClientB["Un-instrumented Client SDK"]
+        ConnectorTypeClientC["Custom client on REST API<br>(i.e. HuggingFaceClient)"]
+    end
+    Connectors["Connectors/AI Services"]
+    blockArrowId2<["&nbsp;&nbsp;&nbsp;"]>(y)
+    SemanticKernel["Semantic Kernel"]
+    block:Kernel
+        Function
+        Planner
+        Agent
+    end
+```
+
+> Semantic Kernel also supports other types of connectors for memories/vector databases. We will discuss instrumentations for those connectors in a separate ADR.
+
+> Note that this will not change our approaches to [instrumentation for planners](./0025-planner-telemetry-enhancement.md). We may modify or remove some of the meters we created previously, which will introduce breaking changes.
+
+In order to keep the activities as close to the model level as possible, we should keep them at or below the connector level.
+
+### Out of scope
+
+These services will be discuss in the future:
+
+- Memory/vector database services
+- Audio to text services (`IAudioToTextService`)
+- Embedding services (`IEmbeddingGenerationService`)
+- Image to text services (`IImageToTextService`)
+- Text to audio services (`ITextToAudioService`)
+- Text to image services (`ITextToImageService`)
+
+## Considered Options
+
+- Instrumentation implementations
+  - Static class
+  - Static member
+  - Dependency injection
+- Where do we create the activities
+  - Add instrumentations at the connector/service level
+  - Add instrumentations at the client level
+
+### Instrumentation implementations
+
+#### Static class `ModelDiagnostics`
+
+This class will live under `dotnet\src\InternalUtilities\src\Diagnostics`.
+
+```C#
+// Example
+namespace Microsoft.SemanticKernel;
+
+internal static class ModelDiagnostics
+{
+    private static string s_namespace = typeof(ModelDiagnostics).Namespace;
+    private static ActivitySource s_activitySource = new ActivitySource(s_namespace);
+
+    public static Activity? StartCompletionActivity(string name, string modelName, string modelProvider, string prompt, PromptExecutionSettings? executionSettings)
+    {
+        ...
+    }
+
+    // Can be used for both non-streaming endpoints and streaming endpoints.
+    // For streaming, collect a list of `StreamingTextContent` and concatenate them into a single `TextContent` at the end of the streaming.
+    void SetCompletionResponses(Activity? activity, IReadOnlyList<TextContent> completions, CompletionUsage? usage, PromptExecutionSettings? executionSettings)
+    {
+        ...
+    }
+
+    ...
+}
+```
+
+Example usage
+
+```C#
+// In client
+public async Task<IReadOnlyList<TextContent>> GenerateTextAsync(
+    string prompt,
+    PromptExecutionSettings? executionSettings,
+    CancellationToken cancellationToken)
+{
+    using var activity = ModelDiagnostics.StartCompletionActivity(
```

Comment:
are those `internal` or client `activities`?

Reply:
These will be internal.